llvm/lib/Target/AMDGPU/SIRegisterInfo.cpp

```diff
@@ -2568,30 +2568,46 @@ bool SIRegisterInfo::eliminateFrameIndex(MachineBasicBlock::iterator MI,
                 } else
                   Add.addImm(Offset).addReg(TmpResultReg, RegState::Kill);
               } else {
-                BuildMI(*MBB, MI, DL, TII->get(AMDGPU::V_MOV_B32_e32),
-                        TmpResultReg)
-                    .addImm(Offset);
                 assert(Offset > 0 &&
                        isUInt<24>(2 * ST.getMaxWaveScratchSize()) &&
                        "offset is unsafe for v_mad_u32_u24");
-                // We start with a frame pointer with a wave space value, and an
-                // offset in lane-space. We are materializing a lane space
-                // value. We can either do a right shift of the frame pointer to
-                // get to lane space, or a left shift of the offset to get to
-                // wavespace. We can right shift after the computation to get
-                // back to the desired per-lane value.
-                // We are using the mad_u32_u24 primarily as an add with no
-                // carry out clobber.
-                Add = BuildMI(*MBB, MI, DL, TII->get(AMDGPU::V_MAD_U32_U24_e64),
+                if (AMDGPU::isInlinableLiteral32(Offset,
+                                                 ST.hasInv2PiInlineImm())) {
```

Comment:
You don't need to duplicate the the paths. You can still use the reassociated path below. You can also conditionally add the operand to the MachineInstrBuilder and skip the materialize

Reply:
it is becoming more ugly to read if i do so,  V_LSHRREV_B32_e64 comes before mad in first case and at the last in second case i have to update MI for them in condition. is it worth over saving few lines.

Reply:
>  V_LSHRREV_B32_e64 comes before mad in first case

It does not need to. Nothing about the order of instructions should change in the inline immediate case. You just avoid the materialize with the v_mov_b32

Reply:
please check once i'm not sure we have to add one more V_LSHRREV_B32_e64 in this case?

Reply:
The only instruction change here is avoiding the initial v_mov_b32. The shifts do not change.

You've moved the code around so now you are emitting instructions out of order. This is harder to follow. Keep the BuildMI for the v_mov_b32 where it was before, just conditional

Reply:
we can only reuse add here in both blocks. not shift. 
first block mad uses shifted value in vgpr. while in other materialized value.

Reply:
I'm thinking to handle mad inside getAddNoCarry now, for that i might require more work.
earlier while i tried to do so got couple of test failures.

Reply:
You're still changing too much and overcomplicating this. This should only change the one path where v_mad_u32_24 is used. You do not need to touch any of the shift handling

Reply:
it is not the case now, that i have handled inside getAddNoCarry

Reply:
This is further complicating a direct fold. This patch should be really small and simple, and not need to touch the helper function 

```
if (!IsInlineImm)
  BuildMI(V_MOV_B32, OffsetReg)

Mad24 = BuildMI(V_MAD_U24_U32, ...)
if (!IsInlineImm)
  Mad24.addReg(OffsetReg)
else
   Mad24.addImm(Imm)...
```